ICE: `in visit, at rust/typecheck/rust-hir-type-check-pattern.cc:`

<!--
File: /tmp/im/c.rs
-->

auto-reduced (treereduce-rust):
````rust
//@compile-flags: -frust-incomplete-and-experimental-compiler-do-not-use
enum A {
    StructWithFields { x: () },
}

fn main() {
    let x = A(3);
    if let A(3) = x {}
}

````

original:
````rust
enum A {
    StructWithFields { x: () },
}


fn main() {
    let x = A(3);
    if let A(3) = x {}
}

````

Version information:

5c3e2adc695b04a199d37cacd8c23b14fdab914f

Possibly related line of code:
https://github.com/Rust-GCC/gccrs/blob/5c3e2adc695b04a199d37cacd8c23b14fdab914f/gcc/rust/typecheck/rust-hir-type-check-pattern.cc#L178-L190

Command:
`/home/matthias/vcs/github/gccrs/gccrs-build/gcc/crab1 -frust-incomplete-and-experimental-compiler-do-not-use`

<details><summary><strong>Program output</strong></summary>
<p>

```
/tmp/icemaker_global_tempdir.kiyYbh4mCz4t/rustc_testrunner_tmpdir_reporting.grGxcOtdDqrC/mvce.rs:6:13: error: expected function, tuple struct or tuple variant, found enum [E0423]
    6 |     let x = A(3);
      |             ^
/tmp/icemaker_global_tempdir.kiyYbh4mCz4t/rustc_testrunner_tmpdir_reporting.grGxcOtdDqrC/mvce.rs:7:19: error: cannot find value ‘x’ in this scope [E0425]
    7 |     if let A(3) = x {}
      |                   ^
crab1: internal compiler error: in visit, at rust/typecheck/rust-hir-type-check-pattern.cc:184
0x2fec938 internal_error(char const*, ...)
	../../gcc/diagnostic-global-context.cc:517
0xc0dfbb fancy_abort(char const*, int, char const*)
	../../gcc/diagnostic.cc:1803
0x90482d Rust::Resolver::TypeCheckPattern::visit(Rust::HIR::TupleStructPattern&)
	../../gcc/rust/typecheck/rust-hir-type-check-pattern.cc:184
0x1094fcf Rust::Resolver::TypeCheckPattern::Resolve(Rust::HIR::Pattern&, Rust::TyTy::BaseType*)
	../../gcc/rust/typecheck/rust-hir-type-check-pattern.cc:39
0x109b794 Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::MatchExpr&)
	../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:1691
0x109a314 Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr&)
	../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:50
0x109bddd Rust::Resolver::TypeCheckExpr::visit(Rust::HIR::BlockExpr&)
	../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:627
0x109a314 Rust::Resolver::TypeCheckExpr::Resolve(Rust::HIR::Expr&)
	../../gcc/rust/typecheck/rust-hir-type-check-expr.cc:50
0x1077370 Rust::Resolver::TypeCheckItem::visit(Rust::HIR::Function&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:642
0x106e105 Rust::Resolver::TypeCheckItem::Resolve(Rust::HIR::Item&)
	../../gcc/rust/typecheck/rust-hir-type-check-item.cc:59
0x100ad4b Rust::Resolver::TypeResolution::Resolve(Rust::HIR::Crate&)
	../../gcc/rust/typecheck/rust-hir-type-check.cc:74
0xdb9778 Rust::Session::compile_crate(char const*)
	../../gcc/rust/rust-session-manager.cc:685
Please submit a full bug report, with preprocessed source (by using -freport-bug).
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.

```

</p>
</details>

--- a/gcc/rust/typecheck/rust-hir-type-check-pattern.cc
+++ b/gcc/rust/typecheck/rust-hir-type-check-pattern.cc
@@ -162,19 +162,32 @@ TypeCheckPattern::visit (HIR::TupleStructPattern &pattern)
 
   infered = pattern_ty;
   TyTy::ADTType *adt = static_cast<TyTy::ADTType *> (infered);
-  rust_assert (adt->number_of_variants () > 0);
 
-  TyTy::VariantDef *variant = adt->get_variants ().at (0);
+  TyTy::VariantDef *variant = nullptr;
   if (adt->is_enum ())
     {
       HirId variant_id = UNKNOWN_HIRID;
       bool ok = context->lookup_variant_definition (
 	pattern.get_path ().get_mappings ().get_hirid (), &variant_id);
-      rust_assert (ok);
+      if (!ok)
+	{
+	  rust_error_at (
+	    pattern.get_locus (), ErrorCode::E0532,
+	    "expected tuple struct or tuple variant, found enum %qs",
+	    pattern_ty->get_name ().c_str ());
+	  return;
+	}
 
       ok = adt->lookup_variant_by_id (variant_id, &variant);
       rust_assert (ok);
     }
+  else
+    {
+      rust_assert (adt->number_of_variants () > 0);
+      variant = adt->get_variants ().at (0);
+    }
+
+  rust_assert (variant != nullptr);
 
   // error[E0532]: expected tuple struct or tuple variant, found struct
   // variant `Foo::D`, E0532 by rustc 1.49.0 , E0164 by rustc 1.71.0

--- a/gcc/testsuite/rust/compile/match-tuplestructpattern-non-variant.rs
+++ b/gcc/testsuite/rust/compile/match-tuplestructpattern-non-variant.rs
@@ -1,0 +1,20 @@
+enum Empty {}
+enum NonEmpty {
+    Foo(i32),
+}
+
+fn f(e: Empty) {
+    match e { 
+        Empty(0) => {} // { dg-error "expected tuple struct or tuple variant, found enum 'Empty'" }
+    }
+
+    match e {
+        Empty(Empty(..)) => {} // { dg-error "expected tuple struct or tuple variant, found enum 'Empty'" }
+    }
+}
+
+fn g(e: NonEmpty) {
+    match e {
+        NonEmpty(0) => {} // { dg-error "expected tuple struct or tuple variant, found enum 'NonEmpty'" }
+    }
+}

